API-Specification.md

```diff
@@ -1,19 +1,264 @@
+
+Relevant terminology from the `HTTP/1.1` (`RFC2616`) standard is used where
+appropriate, such as `header`, `body`.
+
+## Current API Version
+
+By default, all requests to `https://api.uclcssa.cn` SHALL dispatch to the `v1`
+REST API. All requests SHOULD explicitly include the `Accept` HTTP header for
+stability against API changes.
+
+```
+Accept: application/vnd.uclcssa.v1+json
+```
+
+For more details, refer to [Media Types](#media-types).
+
+## Schema
+
+All API communication MUST occur over HTTPS, at `https://api.uclcssa.cn`. The
+data interchange format used MUST be JSON.
+
+For example, the response to a request to `https://api.uclcssa.cn/users/testuser`
+fetching headers should return
+
+```http
+curl -i https://api.uclccsa.cn/users/testuser
+HTTP/1.1 200 OK
+Server: nginx
+Date: Thu, 13 Jun 2019 16:50:11 GMT
+Content-Type: application/json; charset=utf-8
+Content-Encoding: gzip
+Content-Length: 5
+Connection: keep-alive
+Status: 200 OK
+X-Media-Type: uclcssa.v1
+X-Content-Type-Options: nosniff
+```
+
+Suitable HTTP headers SHOULD be used to convey additional information, such as
+authentication, caching, API versioning and access control. See [HTTP Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers) for more information.
+
+Notes:
+- `Date` format SHOULD conform to `HTTP-date` in [RFC7231](https://tools.ietf.org/html/rfc7231)
+- `Content-Encoding` SHOULD be specified to use `gzip`.
+- `X-*` headers are custom headers for additional metadata.
+- `X-Media-Type: uclcssa.v1` specifies the API version as `v1`.
+- There are no rate limits yet - restricting traffic flow, e.g. for
+load-balancing is not yet considered in `v1` API.
+- There are no caching support yet.
+- `Content-Length` specifies the size of the body in bytes, and MUST be
+included.
+
+Blank fields MUST be included with value `null` instead of being omitted.
+
+All timestamps MUST be returned in [RFC3339](https://tools.ietf.org/html/rfc3339)
+format.
+
+## Summary Representations
+
+When requesting a *list* of certain resources, only their *summary
+representations* are returned for performance reasons.
+
+For example, when trying to fetch a list of nearby restaurants, only the summary
+representations of the restaurants are returned.
+
+## Detailed Representations
+
+When requesting an individual resource, such as a `library`, all of its
+attributes and detailed information are returned, depending on the requester's
+permissions.
+
+## Authentication
+
+All authentication for restricting access to server resources MUST be handled
+via [OAuth 2.0](https://oauth.net/2/) mechanisms (hereafter abbreviated as
+`OAuth2`), standardized by
+[RFC6749](http://tools.ietf.org/html/rfc6749).
+
+Tutorials for `OAuth2` can be found at [bubblecode](http://www.bubblecode.net/en/2016/01/22/understanding-oauth2/).
+
+When clients successfully register with the authentication server via
+`UCL email address` and `WeChat id`, they are granted specific OAuth2 `tokens`.
```

Comment:
We should use WeChat Login as our primary Authentication, and UCL OAuth as optional. [More info about WeChat Auth](https://developers.weixin.qq.com/miniprogram/dev/framework/open-ability/login.html)

Reply:
But would UCL OAuth be *optional* though? How would you obtain personal timetables and such without binding to a specific UCL email address? I understand that some functionalities don't need UCL email authentication necessarily, which means stuff like personal timetable sections is gonna need to be hidden in the summary pages in the UI.

Reply:
Yes, as some user might not want to bind their UCL account, we can't force them to bind UCL to use the app. Actually the only thing that needs UCL OAuth is timetable. We can probably simply display a "please bind UCL account" banner in the summary page?

Reply:
I don't think such banner is really necessary, just hide the UCL-specific content unless they explicitly bind it in the MyUCL page or something.